x/farming/legacy/v2/store.go

```diff
@@ -0,0 +1,49 @@
+package v2
+
+import (
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	v1farming "github.com/cosmosquad-labs/squad/x/farming/legacy/v1"
+	"github.com/cosmosquad-labs/squad/x/farming/types"
+)
+
+func MigrateQueuedStaking(store sdk.KVStore, endTime time.Time) {
+	oldStore := prefix.NewStore(store, v1farming.QueuedStakingKeyPrefix)
+
+	oldStoreIter := oldStore.Iterator(nil, nil)
+	defer oldStoreIter.Close()
+
+	for ; oldStoreIter.Valid(); oldStoreIter.Next() {
+		fullKey := append(v1farming.QueuedStakingKeyPrefix, oldStoreIter.Key()...)
+		stakingCoinDenom, farmerAcc := v1farming.ParseQueuedStakingKey(fullKey)
+		store.Set(types.GetQueuedStakingKey(endTime, stakingCoinDenom, farmerAcc), oldStoreIter.Value())
+		oldStore.Delete(oldStoreIter.Key())
+	}
+}
+
+func MigrateQueuedStakingIndex(store sdk.KVStore, endTime time.Time) {
+	oldStore := prefix.NewStore(store, v1farming.QueuedStakingIndexKeyPrefix)
+
+	oldStoreIter := oldStore.Iterator(nil, nil)
+	defer oldStoreIter.Close()
+
+	for ; oldStoreIter.Valid(); oldStoreIter.Next() {
+		fullKey := append(v1farming.QueuedStakingIndexKeyPrefix, oldStoreIter.Key()...)
+		farmerAcc, stakingCoinDenom := v1farming.ParseQueuedStakingIndexKey(fullKey)
+		store.Set(types.GetQueuedStakingIndexKey(farmerAcc, stakingCoinDenom, endTime), oldStoreIter.Value())
+		oldStore.Delete(oldStoreIter.Key())
+	}
+}
+
+func MigrateStore(ctx sdk.Context, storeKey sdk.StoreKey, currentEpochDays uint32) error {
+	store := ctx.KVStore(storeKey)
+
+	endTime := ctx.BlockTime().Add(time.Duration(currentEpochDays) * types.Day)
+	MigrateQueuedStaking(store, endTime)
+	MigrateQueuedStakingIndex(store, endTime)
+
+	return nil
+}
```

Comment:
`endTime` will be adding 24 hours from the current time. Since v1 doesn't keep in track of stakers' queue stakings in time-based, am I understanding correctly that some stakers need to wait longer than they were supposed to wait?

Reply:
No, suppose we have a farmer who has staked their coins at `2022-04-04T21:00:00`. Without the change in queued staking logic, the farmer's coins would be staked at `2022-04-05T00:00:00` and the farmer would receive rewards at `2022-04-06T00:00:00`.
Now, we decided to upgrade the chain and with this migration code, we set `endTime` as 24 hours after the current block time.
So the farmer's queued staking would have `2022-04-05T21:00:00` as `endTime`.
Therefore, at `2022-04-05T21:00:00` the farmer's coins would be staked and the farmer receives rewards at `2022-04-06T00:00:00`, which timing is same as the old logic.

Reply:
It makes sense now. Thanks for explaining with an example. That is helpful.